arroyo/processing/strategies/decoder/base.py

```diff
@@ -23,28 +33,43 @@ class KafkaMessageDecoder(ProcessingStrategy[KafkaPayload], Generic[T]):
     Decode messages to be forwarded to the next step. Optional validation.
     This strategy accepts a KafkaPayload and only performs validation on the
     message value. Headers and keys are forwarded without being validated or parsed.
+
+    Raises ``InvalidMessage`` if the message fails validation.
     """
 
     def __init__(
         self,
         codec: Codec[T],
         validate: bool,
-        next_step: ProcessingStrategy[DecodedKafkaMessage[T]],
+        next_step: ProcessingStrategy[Union[FilteredPayload, DecodedKafkaMessage[T]]],
     ) -> None:
         self.__codec = codec
         self.__validate = validate
         self.__next_step = next_step
+        self.__invalid_messages: MutableSequence[InvalidMessage] = []
+
+    def __forward_invalid_offsets(self) -> None:
+        if self.__invalid_messages:
+            committable = {m.partition: m.offset + 1 for m in self.__invalid_messages}
```

Comment:
this very implicitly relies on the order of messages in `self.__invalid_messages`, particulary when two messages in that array have the same offset. if they are not submitted in the right order, we get this:

```python
{partition: offset + 1 for partition, offset in [(1, 2), (1, 1)]}
# {1: 1}
```

can we at least add some assertions here? I feel that we shouldn't allow misbehaving upstream strategies to have system-wide effects like committing the wrong offsets.

also i think we can hold committable on `self` instead of an entire array of individual `partition, offset` pairs, which i presume reduces memory consumption by quite a bit